File: packages/sdk/src/resource-as.ts
import { randomUUID } from "node:crypto";
import type { JWK } from "jose";
import { isSha256JwkThumbprint } from "./crypto.js";
import { WeldallAuthError } from "./errors.js";
import { hasVerifiedEmail } from "./identity.js";
import { signEs256, verifyEs256 } from "./jwt.js";
import { parseScope } from "./scope.js";
export async function issueAccessToken(input: {
  issuer: string;
  subject: string;
  email: string;
  resource: string;
  clientId: string;
  scopes: string[];
  jkt: string;
  kid: string;
  privateJwk: JWK;
  now?: number;
}): Promise<string> {
  const now = input.now ?? Math.floor(Date.now() / 1000);
  const scope = [...new Set(input.scopes)].sort().join(" ");
  if (
    !input.subject ||
    !hasVerifiedEmail({ email: input.email, email_verified: true }) ||
    !isSha256JwkThumbprint(input.jkt) ||
    !parseScope(scope)
  )
    throw new Error("invalid access-token issuance claims");
  return signEs256(
    {
      iss: input.issuer,
      sub: input.subject,
      email: input.email,
      email_verified: true,
      aud: input.resource,
      client_id: input.clientId,
      scope,
      cnf: { jkt: input.jkt },
      jti: randomUUID(),
      iat: now,
      exp: now + 600,
    },
    { ...input, typ: "at+jwt" },
  );
}
export async function verifyAccessToken(
  token: string,
  input: {
    issuer: string;
    resource: string;
    kid: string;
    publicJwk: JWK;
    clientId: string;
    requiredScopes?: readonly string[];
  },
) {
  const p = await verifyEs256(token, {
    ...input,
    audience: input.resource,
    typ: "at+jwt",
    maxTokenAge: "10m",
    errorCode: "invalid_token",
    errorStatus: 401,
  });
  const cnf = p.cnf;
  const granted = parseScope(p.scope);
  if (
    p.aud !== input.resource ||
    !cnf ||
    typeof cnf !== "object" ||
    Array.isArray(cnf) ||
    !isSha256JwkThumbprint((cnf as { jkt?: unknown }).jkt) ||
    typeof p.sub !== "string" ||
    !p.sub ||
    !hasVerifiedEmail(p) ||
    p.client_id !== input.clientId ||
    typeof p.jti !== "string" ||
    p.jti.length < 1 ||
    p.jti.length > 128 ||
    !Number.isInteger(p.iat) ||
    !Number.isInteger(p.exp) ||
    (p.exp as number) <= (p.iat as number) ||
    (p.exp as number) - (p.iat as number) > 600 ||
    !granted
  )
    throw new WeldallAuthError("invalid_token", "invalid access token claims", 401);
  if (input.requiredScopes?.some((scope) => !granted.includes(scope)))
    throw new WeldallAuthError(
      "insufficient_scope",
      "required scope is missing",
      403,
      input.requiredScopes,
    );
  return p as typeof p & {
    cnf: { jkt: string };
    sub: string;
    email: string;
    email_verified: true;
    scope: string;
  };
}

File: packages/python-sdk/src/weldall/resource_as.py
"""Resource authorization-server access-token helpers."""

from __future__ import annotations

import time
import uuid
from collections.abc import Mapping, Sequence
from typing import Any, NotRequired, TypedDict, overload

from .crypto import is_sha256_jwk_thumbprint
from .errors import WeldallAuthError
from .identity import has_verified_email
from .jwt import sign_es256, verify_es256
from .scope import parse_scope
from .types import JWK, JWTPayload


class IssueAccessTokenInput(TypedDict):
    issuer: str
    subject: str
    email: str
    resource: str
    client_id: str
    scopes: Sequence[str]
    jkt: str
    kid: str
    private_jwk: JWK
    now: NotRequired[int]


class VerifyAccessTokenInput(TypedDict):
    kid: str
    public_jwk: JWK
    issuer: str
    resource: str
    client_id: str
    required_scopes: NotRequired[Sequence[str]]


@overload
def issue_access_token(input: IssueAccessTokenInput, /) -> str: ...


@overload
def issue_access_token(
    *,
    issuer: str,
    subject: str,
    email: str,
    resource: str,
    client_id: str,
    scopes: Sequence[str],
    jkt: str,
    kid: str,
    private_jwk: JWK,
    now: int | None = None,
) -> str: ...


def issue_access_token(input: Mapping[str, Any] | None = None, **kwargs: Any) -> str:
    values = dict(input or {})
    values.update(kwargs)
    now_value = values.get("now")
    now = int(time.time()) if now_value is None else now_value
    scope = " ".join(sorted(set(values["scopes"])))
    if (
        not values["subject"]
        or not has_verified_email({"email": values["email"], "email_verified": True})
        or not is_sha256_jwk_thumbprint(values["jkt"])
        or parse_scope(scope) is None
    ):
        raise ValueError("invalid access-token issuance claims")
    return sign_es256(
        {
            "iss": values["issuer"],
            "sub": values["subject"],
            "email": values["email"],
            "email_verified": True,
            "aud": values["resource"],
            "client_id": values["client_id"],
            "scope": scope,
            "cnf": {"jkt": values["jkt"]},
            "jti": str(uuid.uuid4()),
            "iat": now,
            "exp": now + 600,
        },
        kid=values["kid"],
        private_jwk=values["private_jwk"],
        typ="at+jwt",
    )


@overload
def verify_access_token(token: str, input: VerifyAccessTokenInput, /) -> JWTPayload: ...


@overload
def verify_access_token(
    token: str,
    *,
    kid: str,
    public_jwk: JWK,
    issuer: str,
    resource: str,
    client_id: str,
    required_scopes: Sequence[str] | None = None,
) -> JWTPayload: ...


def verify_access_token(
    token: str, input: Mapping[str, Any] | None = None, **kwargs: Any
) -> JWTPayload:
    values = dict(input or {})
    values.update(kwargs)
    payload = verify_es256(
        token,
        kid=values["kid"],
        public_jwk=values["public_jwk"],
        issuer=values["issuer"],
        audience=values["resource"],
        typ="at+jwt",
        max_token_age_s=600,
        error_code="invalid_token",
        error_status=401,
    )
    cnf = payload.get("cnf")
    granted = parse_scope(payload.get("scope"))
    if (
        payload.get("aud") != values["resource"]
        or not isinstance(cnf, dict)
        or not is_sha256_jwk_thumbprint(cnf.get("jkt"))
        or not isinstance(payload.get("sub"), str)
        or not payload["sub"]
        or not has_verified_email(payload)
        or payload.get("client_id") != values["client_id"]
        or not isinstance(payload.get("jti"), str)
        or not 1 <= len(payload["jti"]) <= 128
        or type(payload.get("iat")) is not int
        or type(payload.get("exp")) is not int
        or payload["exp"] <= payload["iat"]
        or payload["exp"] - payload["iat"] > 600
        or granted is None
    ):
        raise WeldallAuthError("invalid_token", "invalid access token claims", 401)
    required: Sequence[str] | None = values.get("required_scopes")
    if required and any(scope not in granted for scope in required):
        raise WeldallAuthError(
            "insufficient_scope", "required scope is missing", 403, list(required)
        )
    return payload
